### PR TITLE
Trying Out DSL for Go

### PR DESCRIPTION
**Pull request by xiazcy**
Needed to export Bytecode field in graph traversal and some Bytecode methods for creating DSL. The dsl package is for review purposes, and will be put into the documentation file and removed from the actual PR.

Please give this a try locally with the docker server.

### CODE REVIEW

**Comment by xiazcy** [comment]
Please comment if there are other better way of embedding/extending the structs. Link to python DSL for reference (other languages are above) https://tinkerpop.apache.org/docs/3.5.3-SNAPSHOT/reference/#gremlin-python-dsl

File: gremlin-go/dsl/go.mod [hunk 1]
```diff
@@ -0,0 +1,31 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (simonz-bq) [inline gremlin-go/dsl/go.mod]: Note: this will likely mess up releases if it's committed.

File: gremlin-go/driver/graphBinary.go [hunk 2]
```diff
@@ -718,6 +718,8 @@ func (serializer *graphBinaryTypeSerializer) getType(val interface{}) (dataType,
 			// We can write an array or slice into the list dataType.
 			return listType, nil
 		default:
+			fmt.Println("DSL??")
```
Comment (simonz-bq) [inline gremlin-go/driver/graphBinary.go]: Note: Don't forget to remove this.

File: gremlin-go/driver/bytecode.go [hunk 3]
```diff
@@ -30,7 +30,7 @@ type bytecode struct {
 	bindings           map[string]interface{}
 }
 
-func newBytecode(bc *bytecode) *bytecode {
+func NewBytecode(bc *bytecode) *bytecode {
```
Comment (simonz-bq) [inline gremlin-go/driver/bytecode.go]: I worry about this change. It looks like you're exporting it now to use with NewGraphTraversal?

Is there a better way of doing this? Returning unexported types is not recommended and can cause issues. One of which is that documentation for it cannot be generated, as well as making usage of the method esoteric.

File: gremlin-go/driver/graphBinary.go (hunk 2 above)
Comment (lyndonbauto) [reply gremlin-go/driver/graphBinary.go]: @simonz-bq Someone may want to know `DSL??` though...

Comment (xiazcy) [reply gremlin-go/driver/graphBinary.go]: LOL, yes sorry that was for debug, will remove

File: gremlin-go/dsl/go.mod (hunk 1 above)
Comment (xiazcy) [reply gremlin-go/dsl/go.mod]: Ack, and yes this whole DSL package will not be committed.

File: gremlin-go/driver/bytecode.go (hunk 3 above)
Comment (xiazcy) [reply gremlin-go/driver/bytecode.go]: Yes, it is so we can extend the anonymous traversal functions to add steps. The start of each anonymous traversal needs a new bytecode object, so I'm not sure if there is a way to use it without exporting this...

Comment (xiazcy) [reply gremlin-go/driver/bytecode.go]: In any case, I'll see if there are other way to wrap this without exporting it. 
Do you have any comments regarding the other exports or is this one the only concern?

**Review by vkagamlyk — APPROVED** [review]
great job :)

**Pull request closed without merging**